src/platform/datapath_raw_socket.c

```diff
@@ -342,12 +342,77 @@ CxPlatDpRawCalculateHeaderBackFill(
     return HeaderBackFill;
 }
 
+_IRQL_requires_max_(DISPATCH_LEVEL)
+uint16_t
+CxPlatFramingChecksum(
+    _In_reads_(Length) uint8_t* Data,
+    _In_ uint32_t Length,
+    _In_ uint64_t InitialChecksum
+    )
+{
+    //
+    // Checksum is calculated in reverse order.
+    // 1. Add the odd byte to the checksum if the length is odd.
+    // 2. If the length is divided by 2 but not 4, add the last 2 bytes.
+    // 3. Sum up the rest as 32-bit words.
```

Comment:
Folding on each iteration is what I was trying to avoid. Without using assembly intrinsic, we would need to use 128-bit words to store the intermediate result, which I don't know if it would be faster or not.